### ISSUE TITLE
Hardware Float16 on A64fx

### ISSUE THREAD

**Issue by milankl** [JuliaLang/julia#40216]
On arm's A64fx (i.e. aarch64) the official release (Julia 1.6) converts Float16s to Float32 for arithmetics, although A64fx natively supports Float16
```julia
julia> a,b = rand(Float16,2);

julia> @code_llvm a+b
;  @ float.jl:324 within `+'
define half @"julia_+_174"(half %0, half %1) {
top:
  %2 = fpext half %0 to float
  %3 = fpext half %1 to float
  %4 = fadd float %2, %3
  %5 = fptrunc float %4 to half
  ret half %5
}
```
So @vchuravy suggest to comment out https://github.com/JuliaLang/julia/blob/91c297b6c983aed2828600bcd9b8ba6494f747b6/src/aotcompile.cpp#L622-L623
and build 1.6 from source. This worked fine, and now Float16s are not converted
```julia
julia> @code_llvm a+b
;  @ float.jl:324 within `+'
define half @"julia_+_174"(half %0, half %1) {
top:
  %2 = fadd half %0, %1
  ret half %2
}
```
and give a typical 4x speed-up on arithmetic operations compared to Float64. Unless subnormals are hit, see #40151

**Comment by milankl** [comment]
`make testall` of this build yields 2 errors, however, they are not related to Float16, so @vchuravy suggested to include a check for A64fx architecture to address this automatically in the future

```julia
Test Summary:                       |     Pass  Fail  Error  Broken     Total
  Overall                           | 38937450     1      1  352594  39290046
    unicode/utf8                    |       19                             19
    ...
    file                            |     1035     1                     1036
    ...
    Distributed                     |                     1                 1

The global RNG seed was 0x392db231c37d2b35344a0bbd2d5677dc.

Error in testset file:
Test Failed at /snx11273/home/ri-mkloewer/exjul/build/julia-1.6.0/test/file.jl:98
  Expression: dirname(t) == d
   Evaluated: "/var/tmp/pbs.4045.a64fx00" == "/var/tmp/pbs.4045.a64fx00/jl_7cuVre"
Error in testset Distributed:
Error During Test at /snx11273/home/ri-mkloewer/exjul/build/julia-1.6.0/test/testdefs.jl:21
  Got exception outside of a @test
  LoadError: Distributed test failed, cmd : `/snx11273/home/ri-mkloewer/exjul/build/julia-1.6.0/usr/bin/julia
-Cnative -J/snx11273/home/ri-mkloewer/exjul/build/julia-1.6.0/usr/lib/julia/sys.so -g1 --startup-file=no
--check-bounds=yes --startup-file=no --depwarn=error /snx11273/home/ri-mkloewer/exjul/build
/julia-1.6.0/usr/share/julia/stdlib/v1.6/Distributed/test/distributed_exec.jl`
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:33
    [2] top-level scope
      @ /snx11273~/exjul/build/julia-1.6.0/usr/share/julia/stdlib/v1.6/Distributed/test/runtests.jl:11
    ...
  in expression starting at /snx11273/home/ri-mkloewer/exjul/build/julia-1.6.0/usr/share/julia/stdlib/v1.6/Distributed/test/runtests.jl:10

**Comment by vchuravy** [comment]
After chatting with @maleadt today the necessary steps are likely:

1. Check in `demote-float16` pass if the JIT target support Float16 
2. Add support in multi-versioning for splitting functions based on the presence of Float16 operations.

@yuyichao Should one add a new `JL_TARGET_CLONE_FLOAT16`? Or is piggybacking on `JL_TARGET_CLONE_MATH` okay?

**Comment by yuyichao** [comment]
The cloning flag is a pretty rough heuristic mainly because it's hard to model what can be better optimized exactly and not because I didn't want to be more accurate......

In this case, I think it's fairly easy to detect the usage accurately so having a separate cloning flag should be fine. `fp16fml` and `fullfp16` should probably trigger this condition.

On a related note both ARM and x86 now have bf16 extensions and LLVM also has support for it recently so there can probably be a flag for that too.

**Comment by gbaraldi** [comment]
So the solution to this is something similar to #43085 ?

**Comment by maleadt** [comment]
It doesn't need any intrinsics, just conditionally running the pass based on the target machine and integrating with multiversioning.

**Comment by giordano** [comment]
it may be worth mentioning that this would be useful also for the Apple M1 CPUs, which also have support for fp16 and are a bit more accessible than A64FX.

**Comment by gbaraldi** [comment]
For sure, the PR that fixes this is on hold it should be done on a per function basis and you can't query the target machine inside a pass.

**Comment by ctkelley** [comment]
Does @gbaraldi's comment mean that Float16 support won't be in 1.8 for Apple silicon?

**Comment by pengtu** [comment]
Can extend this to add tracking of f16 support for AVX512-fp16 in addition to A64fx?

**Comment by gbaraldi** [comment]
The PR has some preliminary code in order to support it, but I believe a separate one would be necessary. When we add the features for the eventual CPUs that have this it shouldn't be too hard.

**Comment by giordano** [comment]
On c5fe17b821b, which includes #43327, I get the following on A64FX

```llvm
julia> code_llvm(+, (Float16, Float16))
;  @ float.jl:406 within `+`
define half @"julia_+_971"(half %0, half %1) #0 {
top:
  %thread_ptr = call i8* asm "mrs $0, tpidr_el0", "=r"() #2
  %ppgcstack_i8 = getelementptr i8, i8* %thread_ptr, i64 16
  %ppgcstack = bitcast i8* %ppgcstack_i8 to {}****
  %pgcstack = load {}***, {}**** %ppgcstack, align 8
  %ptls_field3 = getelementptr inbounds {}**, {}*** %pgcstack, i64 2
  %2 = bitcast {}*** %ptls_field3 to i64***
  %ptls_load45 = load i64**, i64*** %2, align 8
  fence syncscope("singlethread") seq_cst
  %3 = getelementptr inbounds i64*, i64** %ptls_load45, i64 2
  %safepoint = load i64*, i64** %3, align 8
  %4 = load volatile i64, i64* %safepoint, align 8
  fence syncscope("singlethread") seq_cst
  %5 = fadd half %0, %1
  ret half %5
}
```

No more longer half <-> float back and forth, but there is some extra stuff.  Is that expected?  I see the same with the other data types:
```llvm
julia> code_llvm(+, (Float32, Float32))
;  @ float.jl:406 within `+`
define float @"julia_+_1080"(float %0, float %1) #0 {
top:
  %thread_ptr = call i8* asm "mrs $0, tpidr_el0", "=r"() #1
  %ppgcstack_i8 = getelementptr i8, i8* %thread_ptr, i64 16
  %ppgcstack = bitcast i8* %ppgcstack_i8 to {}****
  %pgcstack = load {}***, {}**** %ppgcstack, align 8
  %ptls_field3 = getelementptr inbounds {}**, {}*** %pgcstack, i64 2
  %2 = bitcast {}*** %ptls_field3 to i64***
  %ptls_load45 = load i64**, i64*** %2, align 8
  fence syncscope("singlethread") seq_cst
  %3 = getelementptr inbounds i64*, i64** %ptls_load45, i64 2
  %safepoint = load i64*, i64** %3, align 8
  %4 = load volatile i64, i64* %safepoint, align 8
  fence syncscope("singlethread") seq_cst
  %5 = fadd float %0, %1
  ret float %5
}

julia> code_llvm(+, (Float64, Float64))
;  @ float.jl:406 within `+`
define double @"julia_+_1082"(double %0, double %1) #0 {
top:
  %thread_ptr = call i8* asm "mrs $0, tpidr_el0", "=r"() #2
  %ppgcstack_i8 = getelementptr i8, i8* %thread_ptr, i64 16
  %ppgcstack = bitcast i8* %ppgcstack_i8 to {}****
  %pgcstack = load {}***, {}**** %ppgcstack, align 8
  %ptls_field3 = getelementptr inbounds {}**, {}*** %pgcstack, i64 2
  %2 = bitcast {}*** %ptls_field3 to i64***
  %ptls_load45 = load i64**, i64*** %2, align 8
  fence syncscope("singlethread") seq_cst
  %3 = getelementptr inbounds i64*, i64** %ptls_load45, i64 2
  %safepoint = load i64*, i64** %3, align 8
  %4 = load volatile i64, i64* %safepoint, align 8
  fence syncscope("singlethread") seq_cst
  %5 = fadd double %0, %1
  ret double %5
}
```

but on x86_64 I see only the sum (and the half<->float conversion for Float16, as expected)
```llvm
julia> code_llvm(+, (Float16, Float16))
;  @ float.jl:383 within `+`
define half @"julia_+_230"(half %0, half %1) #0 {
top:
  %2 = fpext half %0 to float
  %3 = fpext half %1 to float
  %4 = fadd float %2, %3
  %5 = fptrunc float %4 to half
  ret half %5
}

julia> code_llvm(+, (Float32, Float32))
;  @ float.jl:383 within `+`
define float @"julia_+_238"(float %0, float %1) #0 {
top:
  %2 = fadd float %0, %1
  ret float %2
}

julia> code_llvm(+, (Float64, Float64))
;  @ float.jl:383 within `+`
define double @"julia_+_240"(double %0, double %1) #0 {
top:
  %2 = fadd double %0, %1
  ret double %2
}
```

**Comment by giordano** [comment]
I was checking x86_64 on v1.8, but @gbaraldi pointed out offline that the extra code is only on master and it is expected.

So I'd say the original issue is indeed fixed by #43327 :partying_face:  I'm going to close this ticket
